jerry-ext/arg/arg-transform-functions.c

```diff
@@ -84,14 +85,91 @@ jerryx_arg_transform_number (jerryx_arg_js_iterator_t *js_arg_iter_p, /**< avail
                                (jerry_char_t *) "It can not be converted to a number.");
   }
 
-  double *dest = c_arg_p->dest;
-  *dest = jerry_get_number_value (to_number);
+  *number_p = jerry_get_number_value (to_number);
   jerry_release_value (to_number);
 
   return jerry_create_undefined ();
+} /* jerryx_arg_transform_number_common */
+
+/**
+ * clamp a double by given min and max value
+ * @return clamped double value
+ */
+static double
+jerryx_arg_clamp_double (double d, /**< the double value */
+                         double min, /**< min value */
+                         double max) /**< max value */
+{
+  const double t = d < min ? min : d;
+  return t > max ? max : t;
+} /* jerryx_arg_clamp_double */
+
+/**
+ * Tranform a JS argument to a double. Type coercion is not allowed.
+ *
+ * @return jerry undefined: the transformer passes,
+ *         jerry error: the transformer fails.
+ */
+jerry_value_t
+jerryx_arg_transform_number_strict (jerryx_arg_js_iterator_t *js_arg_iter_p, /**< available JS args */
+                                    const jerryx_arg_t *c_arg_p) /**< the native arg */
+{
+  return jerryx_arg_transform_number_strict_common (js_arg_iter_p,
+                                                    c_arg_p->dest);
+} /* jerryx_arg_transform_number_strict */
+
+/**
+ * Tranform a JS argument to a double. Type coercion is allowed.
+ *
+ * @return jerry undefined: the transformer passes,
+ *         jerry error: the transformer fails.
+ */
+jerry_value_t
+jerryx_arg_transform_number (jerryx_arg_js_iterator_t *js_arg_iter_p, /**< available JS args */
+                             const jerryx_arg_t *c_arg_p) /**< the native arg */
+{
+  return jerryx_arg_transform_number_common (js_arg_iter_p,
+                                             c_arg_p->dest);
 } /* jerryx_arg_transform_number */
 
 /**
+ * Use the macro to define two transform functions for int type.
+ * One is non-strict version and the other is the strict version.
+ */
+#define GENERATE_JERRYX_ARG_TRANSFORM_FUNC_FOR_INT(type, min, max) \
+jerry_value_t jerryx_arg_transform_ ## type ## _strict (jerryx_arg_js_iterator_t *js_arg_iter_p, \
+                                                        const jerryx_arg_t *c_arg_p) \
+{ \
+  double tmp; \
+  jerry_value_t rv = jerryx_arg_transform_number_strict_common (js_arg_iter_p, &tmp);\
+  if (!jerry_value_has_error_flag (rv)) \
+  { \
+    tmp = jerryx_arg_clamp_double (tmp, min, max); \
+    *(type ## _t *) c_arg_p->dest = (type ## _t) tmp; \
```

Comment:
I guess some in use cases one may also not want the implicit clamping, but instead throw an error if the input value is out of the possible range.

So, to summarize we've got these "dimensions" when converting numbers:
1. target storage type (width & signed-ness)
2. rounding behavior (floor, round or ceil)
3. out-of-range values: clamping vs throwing Range error

... and then the common dimensions that aren't specific to numbers:
4. optionality
5. strictness

I think it's important to be able to cater for 2 & 3.
But adding 2 & 3 as arguments `jerryx_arg_(u)int...` will make the signature quite long.
Maybe it makes sense to pass a "config" struct and choose default values such that passing `{}` as config is a reasonable thing to do?

```

typedef struct {
  jerryx_arg_coerce_t coercion;
  jerryx_arg_optional_t optionality;
  jerryx_arg_rounding_t rounding;
  jerryx_arg_clipping_t clipping;
} jerryx_arg_number_options_t;

//------

uint8_t d = 0;
jerryx_arg_t mapping[] = {
  // {} == "defaults": optional arg + coerce types + round + clip.
  jerryx_arg_uint8(&d, {})
};

//-------
uint8_t d = 0;
jerryx_arg_t mapping[] = {
  // {} == "defaults", all zero
  jerryx_arg_uint8(&d, (jerryx_arg_number_options_t) {
   .clipping = JERRYX_ARG_NO_CLIPPING,
   .coercion = JERRYX_ARG_NO_COERCE
  });
};

```